pkg/genericcontrolplane/aggregator.go

```diff
@@ -14,287 +14,184 @@ See the License for the specific language governing permissions and
+		return
 	}
 
-	// create controllers for auto-registration
-	apiRegistrationClient, err := apiregistrationclient.NewForConfig(aggregatorConfig.GenericConfig.LoopbackClientConfig)
+	// Discovery for things like core, authentication, authorization, certificates, ...
+	gcpGroups, err := s.GenericControlPlane.GenericAPIServer.DiscoveryGroupManager.Groups(req.Request.Context(), req.Request)
 	if err != nil {
-		return nil, err
+		http.Error(resp.ResponseWriter, fmt.Sprintf("error retrieving generic control plane discovery groups: %v", err), http.StatusInternalServerError)
 	}
-	autoRegistrationController := autoregister.NewAutoRegisterController(aggregatorServer.APIRegistrationInformers.Apiregistration().V1().APIServices(), apiRegistrationClient)
-	apiServices := apiServicesToRegister(delegateAPIServer, autoRegistrationController, id)
-	crdRegistrationController := crdregistration.NewCRDRegistrationController(
-		apiExtensionInformers.Apiextensions().V1().CustomResourceDefinitions(),
-		autoRegistrationController)
-
-	err = aggregatorServer.GenericAPIServer.AddPostStartHook("kube-apiserver-autoregistration", func(context genericapiserver.PostStartHookContext) error {
-		go crdRegistrationController.Run(5, context.StopCh)
-		go func() {
-			// let the CRD controller process the initial set of CRDs before starting the autoregistration controller.
-			// this prevents the autoregistration controller's initial sync from deleting APIServices for CRDs that still exist.
-			// we only need to do this if CRDs are enabled on this server.  We can't use discovery because we are the source for discovery.
-			if aggregatorConfig.GenericConfig.MergedResourceConfig.AnyVersionForGroupEnabled("apiextensions.k8s.io") {
-				crdRegistrationController.WaitForInitialSync()
-			}
-			autoRegistrationController.Run(5, context.StopCh)
-		}()
-		return nil
-	})
+
+	// Discovery for the apiextensions group itself
+	apiextensionsGroups, err := s.CustomResourceDefinitions.GenericAPIServer.DiscoveryGroupManager.Groups(req.Request.Context(), req.Request)
 	if err != nil {
-		return nil, err
+		http.Error(resp.ResponseWriter, fmt.Sprintf("error retrieving apiextensions discovery groups: %v", err), http.StatusInternalServerError)
 	}
 
-	err = aggregatorServer.GenericAPIServer.AddBootSequenceHealthChecks(
-		makeAPIServiceAvailableHealthCheck(
-			"autoregister-completion",
-			apiServices,
-			aggregatorServer.APIRegistrationInformers.Apiregistration().V1().APIServices(),
-		),
-	)
+	// Discovery for all the groups contributed by CRDs
+	crdGroups, err := s.CustomResourceDefinitions.DiscoveryGroupLister.Groups(req.Request.Context(), req.Request)
 	if err != nil {
-		return nil, err
+		http.Error(resp.ResponseWriter, fmt.Sprintf("error retrieving custom resource discovery groups: %v", err), http.StatusInternalServerError)
 	}
 
-	return aggregatorServer, nil
-}
+	combined := make([]metav1.APIGroup, 0, len(gcpGroups)+len(apiextensionsGroups)+len(crdGroups))
```

Comment:
This is a bit curious use of `make()` - why?

Reply:
We know the length in advance, so I thought we'd set it. What were you thinking of for this instead?

Reply:
You also know you're about to fill it, so why set it to 0 and add a capacity?

Reply:
@stevekuznetsov did some Googling and learned that `copy` is more efficient for combining slices. Updated - PTAL.